text/0000-ssr-rehydration.md

```diff
@@ -0,0 +1,299 @@
+---
+title: SSR Rehydration
+status: DRAFTED
+created_at: 2021-08-31
+updated_at: 2021-08-31
+pr: (leave this empty until the PR is created)
+---
+
+# SSR Rehydration
+
+## Summary
+
+Whereas server-side rendering (SSR) allows a web application to be rendered into HTML on the server, rehydration involves receiving that HTML on the client and reusing the resulting DOM during initial client-side rendering (CSR).
+
+This document proposes an approach to supporting SSR rehydration in LWC.
+
+## Problem description
+
+### Motivation
+
+The motivations for SSR are [well documented](https://medium.com/walmartglobaltech/the-benefits-of-server-side-rendering-over-client-side-rendering-5d07ff2cefe8) and can be boiled down to the following two points:
+
+- SSR allows search engines to easily index URLs that would otherwise be rendered on the client as part of a web application. This results in improved [SEO](https://developers.google.com/search/docs/beginner/seo-starter-guide), with all the associated benefits.
+
+- SSR can improve the perceived startup performance of a web application by reducing the time to [First Contentful Paint](https://web.dev/first-contentful-paint/). Optimizing perceived performance can have measurable impact on [conversion rates](https://www.cloudflare.com/learning/performance/more/website-performance-conversion-rates/), [user satisfaction](https://www.ericsson.com/en/press-releases/2016/2/streaming-delays-mentally-taxing-for-smartphone-users-ericsson-mobility-report), and [revenue](https://www.dareboost.com/en/webperf-impacts).
+
+However, rendering on the server can itself introduce performance issues, working against the original intent. Namely, if DOM that was generated from SSR HTML is not preserved during the initial CSR, this will result in potentially expensive DOM manipulation and unwanted visual artifacts.
+
+The aim of rehydration is to avoid the client-side performance issues of a naive SSR implementation, reusing original DOM wherever possible.
+
+### The Happy Path
+
+With regard to SSR, there exists a happy path (Fig 1a) where the rendering output for a given request is identical on both the server and the client; identical VDOM is generated for the entire document in both environments.
+
+Depending upon the actual implementation, exceptions to this happy path include:

```

Comment:
I can think of another exception, where for example, you want to render a dummy component on the server, for example `<router-container></router-container>`. On the client I would expect this to get fully connected and rendered but on the server I'll render a dummy implementation. It's like an expected mismatch.

Reply:
Can you elaborate on how this is distinct from Exception A?  I think, perhaps, the exception description can be improved and that is where the gap is. Essentially, one thing (likely a placeholder of some kind) is rendered on the server and another thing is swapped in on the client.

There are actually two ways this could go from there:

1. You render something different on the client for the first render during rehydration. This results in an "expected mismatch" to use your words, and the patch takes care of swapping in the new content.
2. You render the same thing on the client for the first CSR during rehydration but hook into `connectedCallback` and update your state to render something else asynchronously on a subsequent/early CSR.

That does bring to mind something else worth considering. Should there exist another lifecycle for when a component is attached to pre-existing DOM during rehydration? I'm not yet aware of how lifecycle hooks are widely used by LWC devs, and I wouldn't want to overload `connectedCallback` and break existing apps.

Reply:
>  Can you elaborate on how this is distinct from Exception A?

So there's really three scenarios we're discussing:

1. The DOM isn't the same and it's unexpected and we don't know what to do so we abort.
2. The DOM is intentionally different with the expectation that it's ok because it will rehydrate the web component.
3. The DOM is intentionally different because it branched its component logic by interrogating the render state (client vs server).

I think as long as, in all three scenarios we can recover from the mismatch and resume rehydration at a component level, then there's no distinction that needs to be made.

> Should there exist another lifecycle event for when a component is attached to pre-existing DOM during rehydration?

When the component is attached to pre-existing DOM wouldn't that effectively be the `render` phase? Thus, `renderedCallback` might be suitable?

Reply:
> As far as I understand how rehydration work in other frameworks, the 2 examples above will make the framework bails out and re-render the page from scratch. Minor differences can be recovered, like text change. But DOM structural changes can't be recovered during rehydration because it would require a full DOM diffing.

There is some variation across frameworks for how this is handled and it has changed over time. Historically, for example, React was resilient to mismatches during SSR rehydration. React 16 changed that behavior for performance reasons.

However, this isn't a binary thing (resilient vs not resilient, full DOM diffing vs complete bailtout) and we don't have to treat it that way. When considering how mismatches are handled during rehydration, the following options come to mind:

1. Bail out of rehydration entirely when a mismatch is detected, throw away the original HTML, and force a full re-render.
2. Handle easy mismatches (text nodes, certain attrs, etc.) falling back to number 1 for anything more extreme.
3. Bail out of rehydration for subtrees, rooted at a mismatch.
4. Bail out of rehydration for subtrees, rooted at the component where the mismatch occurred.
5. Try to repair as much of the SSR-derived DOM as possible, throwing nothing away explicitly.
6. Assume no mismatches exist, and break silently.

When you refer to full DOM diffing, I interpret that to mean #5. However, we could choose #4, and may get good mileage out of that approach.

So what I'm thinking is:

- We detect a mismatch
- We determine if that mismatch is easily recoverable (e.g. we have allowances for text-node mismatches)
- If it is not easily recoverable, we determine the closest LWC ancestor to that mismatch (anywhere from the root component in a SPA to a leaf component for a small widget).
- We invalidate the ancestor's subtree, replacing SSR-derived DOM with CSR DOM _for that component subtree only_.

It seems that mismatches in LWC will be either A) component-level mismatch due to non-serializable state, B) ephemeral state (e.g. current time in a text node), or C) something broke unexpectedly and we need to recover.

The approach I'm suggesting would handle case A in a reasonable way most times, without incurring horrible performance degredation. If the state for the root component of a huge tree is non-serializable, the action taken would be be roughly comparable to a full bailout, which is probably the best we can do, anyway.

If we make allowances for "easy" mismatches, case B is handled without any notable cost.

The best approach for case C is difficult to know without seeing how SSR is utilized in the real world. But, worst case, you bail out at the root level. In the best case, you bail out for part of the application, minimizing the performance impact.

Reply:
> > Should there exist another lifecycle event for when a component is attached to pre-existing DOM during rehydration?

> Ideally, developers shouldn't have to differentiate between pure client-side rendering and rehydration. @divmain Do you have a use case in mind?

I have seen developers commonly abuse lifecycle events, so I may be overly cautious here. But let's say a developer wants to create a websocket connection to some backend, and they want that connection attached to an instance of a component.

`connectedCallback` makes sense for that. But in a non-SSR world, a component won't have any child elements when `connectedCallback` is invoked. In an SSR world, it would.

I can't think of an example of where that _should_ matter, but I'm thinking [Hyrum's Law](https://www.hyrumslaw.com/) may apply.

Reply:
Should we include a section on handling mismatches (exception c), including spec-like steps?

> As far as I understand how rehydration work in other frameworks, the two examples above will make the framework bails out and re-render the page from scratch. Minor differences can be recovered, like text change. But DOM structural changes can't be recovered during rehydration because it would require a full DOM diffing.

I like this alternative because we do the checks at dev time, bailing out the entire dom being hydrated. For prod, we could assume everything is correct and just hydrate without any verification, recovering minor differences.

> So what I'm thinking is:
We detect a mismatch
We determine if that mismatch is easily recoverable (e.g. we have allowances for text-node mismatches)
If it is not easily recoverable, we determine the closest LWC ancestor to that mismatch (anywhere from the root component in a SPA to a leaf component for a small widget).
We invalidate the ancestor's subtree, replacing SSR-derived DOM with CSR DOM for that component subtree only.

I also consider this attractive: it allows us to reuse more from SSR DOM. But, If we do it only on dev, it is throwaway code because perf is not as important, so the first approach is as good as this one. The benefit may be when used in prod, but there's a perf hit of "mismatch detection" when compared to the previous solution.

Reply:
Open question: How to recover from a lifecycle error in a component while hydrating? (ex: an error on the connected callback)